reaktive/src/commonMain/kotlin/com/badoo/reaktive/observable/TakeObservable.kt

```diff
@@ -0,0 +1,41 @@
+package com.badoo.reaktive.observable
+
+import com.badoo.reaktive.base.subscribeSafe
+import com.badoo.reaktive.completable.CompletableCallbacks
+import com.badoo.reaktive.disposable.Disposable
+import com.badoo.reaktive.disposable.DisposableWrapper
+import com.badoo.reaktive.utils.atomic.AtomicInt
+
+/**
+ * Emit only the first [limit] items emitted by source.
+ */
+fun <T> Observable<T>.take(limit: Int): Observable<T?> {
+    require(limit >= 0) { "count >= 0 required but it was $limit" }
+
+    return observable { emitter ->
+        val disposableWrapper = DisposableWrapper()
+        emitter.setDisposable(disposableWrapper)
+
+        val remaining = AtomicInt(limit)
+
+        subscribeSafe(object : ObservableObserver<T>, CompletableCallbacks by emitter {
+            override fun onSubscribe(disposable: Disposable) {
+                disposableWrapper.set(disposable)
+
+                if (remaining.value == 0) {
```

Comment:
Don't you think that if limit is 0 then there is no need to subscribe to the upstream at all? Perhaps you can just call `emitter.onComplete()` and return right after `emitter.setDisposable(...)` in this case. RxJava2 actually subscribes and then immediately completes from `onSubscribe`, exactly as you did. But I don't see any points of such a behaviour. Better to just do nothing.

Reply:
Please also make sure there is a test for either case: explicit test that it does not subscribe to the upstream if `limit` is 0.

Reply:
> Don't you think that if limit is 0 then there is no need to subscribe to the upstream at all?

I believe there is an important distinction: the disposal only happens once it's subscribed. I think that has always been the expected behavior for Rx where cogs start moving only when they're subscribed and we shouldn't be making Reaktive incompatible.

Reply:
> Please also make sure there is a test for either case: explicit test that it does not subscribe to the upstream if limit is 0.

This one?
https://github.com/badoo/Reaktive/pull/193/files#diff-00df270188cf11ffd754c95a7e37d345R15

Reply:
You are getting me wrong. If you will not subscribe to the upstream and just complete then you will not violate any rx rules. You don't even need to set a desposable for the emitter. Just call emitter.onComplete() and return from the very beginning.

Reply:
So if you write something like:
```
    observable { emitter ->
        if (limit == 0) {
            emitter.onComplete()
            return@observable
        }
        
        // rest of the code
```

then it should be fine. Downstream observer`s `onSubscribe` will be already called at this point by `observable{}` builder function and you are free to produce any events from the beginning. There is no obligation to subscribe to the upstream.

Reply:
Alright, looks like it is better to follow RxJava in this case.